substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrSymbolRepository.java

```diff
@@ -87,12 +88,36 @@ private long getSymbolId(String imageHeapString, boolean replaceDotWithSlash) {
 
     @Override
     public void write(JfrChunkWriter writer) throws IOException {
+        // TODO: Epoch-based constant pool writing should allow for
+        // writing outside of a safepoint. However symbols can be
+        // marked in use directly by native events so the symbol
+        // repository probably needs to be epoch based as well
         assert VMOperation.isInProgressAtSafepoint();
         writer.writeCompressedLong(JfrTypes.Symbol.getId());
         writer.writeCompressedLong(table.getSize());
 
-        // TODO: iterate the table and write the symbol data in the correct encoding. Also consider
-        // symbol.getReplaceDotWithSlash() when writing the data.
+        JfrSymbol[] entries = table.getTable();
+        for (int i = 0; i < entries.length; i++) {
+            JfrSymbol entry = entries[i];
+            if (entry.isNonNull()) {
+                while (entry.isNonNull()) {
+                    writeSymbol(writer, entry);
+                    entry = entry.getNext();
+                }
+            }
+        }
+    }
+
+    private void writeSymbol(JfrChunkWriter writer, JfrSymbol symbol) throws IOException {
+        writer.writeCompressedLong(symbol.getId());
+        byte utf8 = 3;
+        writer.writeByte(utf8);
+        writer.writeCompressedInt(symbol.getValue().length());
```

Comment:
What is needed here, the number of bytes or the number of characters?

Reply:
`write(const Symbol* symbol)`
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/jfr/writers/jfrWriterHost.inline.hpp#L269
and `write_symbol`
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/jfr/recorder/checkpoint/types/jfrTypeSet.cpp#L984

uses `write_utf8(const char* value)`
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/jfr/writers/jfrWriterHost.inline.hpp#L92

which writes `strlen`. That corresponds to Java `String.length()` right? (while `sizeof` would be number of bytes?)


However the jfr tool `StringParser` does charsets like UTF_8 by bytes so maybe I missed something. I didn't confirm yet if StringParser is used to read symbol values yet though.
https://github.com/openjdk/jdk/blob/master/src/jdk.jfr/share/classes/jdk/jfr/internal/consumer/StringParser.java#L71

Reply:
Thanks for looking that up. Seems that it is necessary to write the number of bytes (`strlen(...)` will return the number of bytes when it is applied to `char*`).

Reply:
Ah my bad; I will switch to number of bytes. Thank you for bringing this up!